datatransferkit@3.1.1 ^trilinos@14.0.0:DTK_Topology.hpp:104:30: 'Basis_HGRAD_QUAD_DEG2_FEM' in namespace 'Intrepid2::Impl' does not name a template type

### Steps to reproduce the issue

`datatransferkit@3.1.1%gcc@11.4.0~external-arborx~ipo~openmp+serial+shared build_system=cmake` build fails with:
* spack/develop (62160021c1475eb5b10f0ff948412fdb78044390 from `Tue Feb 27 13:52:47 2024 -0800`)
* Ubuntu 22.04, GCC 11.4.0, X86_64
* Trilinos 14.0.0

<details><summary>Concretization</summary>

<pre>
==> Concretized datatransferkit
 -   f323ieg  datatransferkit@3.1.1%gcc@11.4.0~external-arborx~ipo~openmp+serial+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
[+]  euqbgfo      ^boost@1.79.0%gcc@11.4.0+atomic+chrono~clanglibcpp+container~context~contract~coroutine+date_time~debug+exception~fiber+filesystem+graph~graph_parallel~icu+iostreams~json+locale+log+math+mpi+multithreaded~nowide~numpy~pic+program_options~python+random+regex+serialization+shared+signals~singlethreaded+stacktrace+system~taggedlayout+test+thread+timer~type_erasure~versionedlayout+wave build_system=generic cxxstd=17 patches=a440f96,b8569d7 visibility=hidden arch=linux-ubuntu22.04-x86_64_v3
[+]  x6xxeso          ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]  jz67dau              ^diffutils@3.9%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  xbtzxbw          ^xz@5.4.6%gcc@11.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]  ak6hpj5          ^zlib-ng@2.1.5%gcc@11.4.0+compat+opt build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  k5ioakp          ^zstd@1.5.5%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]  5tibso2      ^cmake@3.27.9%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-x86_64_v3
[+]  4em5peb          ^curl@8.6.0%gcc@11.4.0~gssapi~ldap+libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-x86_64_v3
[+]  duq24qn              ^libidn2@2.3.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  6lom7d2                  ^libunistring@1.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  tsxoxvv              ^nghttp2@1.57.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  jdnziex              ^openssl@3.2.1%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-x86_64_v3
[+]  orzhzee                  ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]  lvpywf6          ^ncurses@6.4%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  gyvgctv      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]  wtqmyvb      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu22.04-x86_64_v3
[+]  w7snqgj      ^mpich@4.1.2%gcc@11.4.0~argobots~cuda+fortran~hwloc+hydra+libxml2+pci~rocm+romio~slurm~vci~verbs~wrapperrpath build_system=autotools datatype-engine=auto device=ch4 netmod=ofi pmi=pmi arch=linux-ubuntu22.04-x86_64_v3
[+]  6a3ezky          ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu22.04-x86_64_v3
[+]  qbe2tgh          ^libfabric@1.20.1%gcc@11.4.0~debug~kdreg~uring build_system=autotools fabrics=rxm,sockets,tcp,udp arch=linux-ubuntu22.04-x86_64_v3
[+]  m4c77rz          ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  sojgcrc              ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  7f2zoxe              ^util-macros@1.19.3%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  xd27xmc          ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  4bin4z7              ^libiconv@1.17%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]  uzzbkdc          ^pkgconf@1.9.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  js34bvw          ^yaksa@0.3%gcc@11.4.0~cuda~rocm build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  isfgk7s              ^autoconf@2.72%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  qkv54hd              ^automake@1.16.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  uaca3vq              ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-x86_64_v3
[+]  kcyhkht                  ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  optd4on              ^python@3.10.13%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=0d98e93,7d40923,ebdca64,f2fd060 arch=linux-ubuntu22.04-x86_64_v3
[+]  mrupjjx                  ^expat@2.6.0%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  75sfst2                      ^libbsd@0.11.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  mnxtea4                          ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  vyp7lu7                  ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  zckvcsr                  ^gettext@0.22.4%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  3w7w7ki                      ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu22.04-x86_64_v3
[+]  zvyqqaz                          ^pigz@2.8%gcc@11.4.0 build_system=makefile arch=linux-ubuntu22.04-x86_64_v3
[+]  qa4bvzv                  ^libffi@3.4.4%gcc@11.4.0 build_system=autotools patches=070b1f3 arch=linux-ubuntu22.04-x86_64_v3
[+]  oljurle                  ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu22.04-x86_64_v3
[+]  owpxnfh                  ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-x86_64_v3
[+]  lt5gstq                  ^sqlite@3.43.2%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  i3erqqz                  ^util-linux-uuid@2.38.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-x86_64_v3
[+]  rvth32h      ^trilinos@14.0.0%gcc@11.4.0~adelus~adios2+amesos+amesos2+anasazi+aztec~basker+belos+boost~chaco~complex~cuda~cuda_rdc~debug~dtk+epetra+epetraext~epetraextbtf~epetraextexperimental~epetraextgraphreorderings~exodus+explicit_template_instantiation~float+fortran~gtest~hdf5~hypre+ifpack+ifpack2+intrepid+intrepid2~ipo+isorropia+kokkos~mesquite+minitensor+ml+mpi+muelu~mumps+nox~openmp~panzer+phalanx+piro~python~rocm~rocm_rdc+rol+rythmos+sacado~scorec+shards+shared+shylu+stk+stokhos+stratimikos~strumpack~suite-sparse~superlu+superlu-dist+teko+tempus~test+thyra+tpetra+trilinoscouplings~wrapper~x11+zoltan+zoltan2 build_system=cmake build_type=Release cxxstd=17 generator=make gotype=long_long patches=bcd552d arch=linux-ubuntu22.04-x86_64_v3
[+]  ygmyfu3          ^hwloc@2.9.1%gcc@11.4.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu22.04-x86_64_v3
[+]  ze7hpgh          ^metis@5.1.0%gcc@11.4.0~gdb~int64~ipo~real64+shared build_system=cmake build_type=Release generator=make patches=4991da9,93a7903,b1225da arch=linux-ubuntu22.04-x86_64_v3
[+]  e72y242          ^openblas@0.3.26%gcc@11.4.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=openmp arch=linux-ubuntu22.04-x86_64_v3
[+]  3x6v26a              ^perl@5.38.0%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-x86_64_v3
[+]  vpzzlp3                  ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-x86_64_v3
[+]  3oxlbms          ^parmetis@4.0.3%gcc@11.4.0~gdb~int64~ipo+shared build_system=cmake build_type=Release generator=make patches=4f89253,50ed208,704b84f arch=linux-ubuntu22.04-x86_64_v3
[+]  zqhzjvj          ^superlu-dist@8.2.1%gcc@11.4.0~cuda~int64~ipo~openmp+parmetis~rocm+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-x86_64_v3
</pre></details>

Error:
```
...
==> Installing datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a [27/27]
==> No binary for datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/git//ORNL-CEES/DataTransferKit.git/bfb7673cc233c26a6a541cbf096f37f26df1e5fb.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> No patches needed for datatransferkit
==> datatransferkit: Executing phase: 'cmake'
==> datatransferkit: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'
...
     316    In file included from /tmp/root/spack-stage/spack-stage-datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a/spack-src/packages/Discretization/src/DTK_FE.hpp:17,
     317                     from /tmp/root/spack-stage/spack-stage-datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a/spack-src/packages/Discretization/src/DTK_FE.cpp:12:
  >> 318    /tmp/root/spack-stage/spack-stage-datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a/spack-src/packages/Discretization/src/DTK_Topology.hpp:86:30: error: 'Basis_HGRAD_HEX_DEG2_FEM' in namespace 'Intrepid2::Impl' does not name a template type; did you mean 'Basis_HGRAD_HEX_C
            2_FEM'?
     319       86 |     typedef Intrepid2::Impl::Basis_HGRAD_HEX_DEG2_FEM<false> basis_type;
     320          |                              ^~~~~~~~~~~~~~~~~~~~~~~~
     321          |                              Basis_HGRAD_HEX_C2_FEM
  >> 322    /tmp/root/spack-stage/spack-stage-datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a/spack-src/packages/Discretization/src/DTK_Topology.hpp:104:30: error: 'Basis_HGRAD_QUAD_DEG2_FEM' in namespace 'Intrepid2::Impl' does not name a template type; did you mean 'Basis_HGRAD_QUA
            D_C2_FEM'?
     323      104 |     typedef Intrepid2::Impl::Basis_HGRAD_QUAD_DEG2_FEM<false> basis_type;
     324          |                              ^~~~~~~~~~~~~~~~~~~~~~~~~
     325          |                              Basis_HGRAD_QUAD_C2_FEM
...
  >> 483    make[2]: *** [packages/Discretization/src/CMakeFiles/dtk_discretization.dir/build.make:93: packages/Discretization/src/CMakeFiles/dtk_discretization.dir/DTK_PointInCell_SERIAL.cpp.o] Error 1
     484    cd /tmp/root/spack-stage/spack-stage-datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a/spack-build-f323ieg/packages/Benchmarks/HybridTransport/src && /spack/opt/spack/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/cmake-3.27.9-5tibso2rqfev26cwtgtcxvwjqsrmkmo7/bin/cmake -E cmake_sy
            mlink_library libdtk_hybridtransport.so.3.1.0 libdtk_hybridtransport.so.3 libdtk_hybridtransport.so
     485    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a/spack-build-f323ieg'
     486    [ 88%] Built target dtk_hybridtransport
  >> 487    make[2]: *** [packages/Discretization/src/CMakeFiles/dtk_discretization.dir/build.make:107: packages/Discretization/src/CMakeFiles/dtk_discretization.dir/DTK_PointSearch_SERIAL.cpp.o] Error 1
  >> 488    make[2]: *** [packages/Discretization/src/CMakeFiles/dtk_discretization.dir/build.make:121: packages/Discretization/src/CMakeFiles/dtk_discretization.dir/DTK_Interpolation_SERIAL.cpp.o] Error 1
     489    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-datatransferkit-3.1.1-f323iegudooyjegoi2efezp66sbg7k7a/spack-build-f323ieg'
  >> 490    make[1]: *** [CMakeFiles/Makefile2:1217: packages/Discretization/src/CMakeFiles/dtk_discretization.dir/all] Error 2
```


### Error message

<details><summary>Error message</summary>

<pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.22.0.dev0 (78b72381578c05a835b82891f7759ef3e6206a0d)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-cascadelake
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/14858509/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/14858510/spack-build-out.txt)

@Rombur 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/datatransferkit/package.py
+++ b/var/spack/repos/builtin/packages/datatransferkit/package.py
@@ -43,7 +43,7 @@ class Datatransferkit(CMakePackage):
     depends_on("trilinos+openmp", when="+openmp")
     depends_on("trilinos+stratimikos+belos", when="@master")
     depends_on("trilinos@13:13.4.1", when="@3.1-rc2:3.1-rc3")
-    depends_on("trilinos@14:", when="@3.1.0:")
+    depends_on("trilinos@14.2:", when="@3.1.0:")
 
     def cmake_args(self):
         spec = self.spec

